File: src/main/java/com/kopyl/commit/CreateCommitAction.java
package com.kopyl.commit;

import com.atlassian.jira.rest.client.api.domain.Issue;
import com.intellij.dvcs.repo.RepositoryManager;
import com.intellij.openapi.project.Project;
import com.kopyl.commit.configuration.AppSettingsState;
import git4idea.GitLocalBranch;
import git4idea.GitUtil;
import git4idea.repo.GitRepository;
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.vcs.CommitMessageI;
import com.intellij.openapi.vcs.VcsDataKeys;
import com.intellij.openapi.vcs.ui.Refreshable;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Objects;

public class CreateCommitAction extends AnAction {

    @Override
    public void actionPerformed(@NotNull AnActionEvent actionEvent) {
        CommitMessageI commitPanel = getCommitPanel(actionEvent);
        if (commitPanel == null) return;

        String personalAccessToken = AppSettingsState.getInstance().personalAccessToken;
        String jiraUrl = AppSettingsState.getInstance().jiraUrl;

        if (Objects.equals(jiraUrl, "") || Objects.equals(personalAccessToken, "")) {
            String errorMessage = "Provide a Jira URL and personal access token in the Settings.";
            ErrorNotification.show(errorMessage);
            return;
        }

        JiraClient jiraClient = new JiraClient(personalAccessToken, jiraUrl);

        try {
            String jiraId = getJiraIdFromBranchName(actionEvent.getProject());
            Issue issue = jiraClient.getIssue(jiraId);

            IssueValidator issueValidator = new IssueValidator(issue);
            if (!issueValidator.valid()) {
                String errorMessage = String.format("Issue %s is not valid", jiraId);
                ErrorNotification.show(errorMessage);
                return;
            }

            CommitMessage commitMessage = new CommitMessage(issue);

            commitPanel.setCommitMessage(commitMessage.toString());
        } catch (Exception e) {
            String errorMessage = String.format("Could not connect to Jira or issue does not exist. Check if %s is available and the access token is valid.", jiraUrl);
            ErrorNotification.show(errorMessage);
        }
    }

    @Nullable
    private static CommitMessageI getCommitPanel(@Nullable AnActionEvent e) {
        if (e == null) {
            return null;
        }
        Refreshable data = Refreshable.PANEL_KEY.getData(e.getDataContext());
        if (data instanceof CommitMessageI) {
            return (CommitMessageI) data;
        }
        return VcsDataKeys.COMMIT_MESSAGE_CONTROL.getData(e.getDataContext());
    }

    private String getJiraIdFromBranchName(Project project) {
        RepositoryManager<GitRepository> repositoryManager = GitUtil.getRepositoryManager(project);
        GitLocalBranch branch = repositoryManager.getRepositories().get(0).getCurrentBranch();

        if (branch == null) return "";

        return branch.getName();
    }
}

File: src/main/kotlin/com/kopyl/commit/CreateCommitAction.kt
package com.kopyl.commit

import com.intellij.dvcs.repo.RepositoryManager
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.project.Project
import com.intellij.openapi.vcs.CommitMessageI
import com.intellij.openapi.vcs.VcsDataKeys
import com.intellij.openapi.vcs.ui.Refreshable
import com.kopyl.commit.configuration.AppSettingsState.Companion.instance
import git4idea.GitUtil
import git4idea.repo.GitRepository

class CreateCommitAction : AnAction() {
    override fun actionPerformed(actionEvent: AnActionEvent) {
        val commitPanel = getCommitPanel(actionEvent) ?: return

        val personalAccessToken = instance.personalAccessToken
        val jiraUrl = instance.jiraUrl
        if (jiraUrl == "" || personalAccessToken == "") {
            val errorMessage = "Provide a Jira URL and personal access token in the Settings."
            ErrorNotification.show(errorMessage)
            return
        }
        val jiraClient = JiraClient(personalAccessToken, jiraUrl)
        try {
            val project = actionEvent.project
            if (project == null) {
                val errorMessage = "No project open."
                ErrorNotification.show(errorMessage)
                return
            }

            val jiraId = getJiraIdFromBranchName(project)
            val issue = jiraClient.getIssue(jiraId)
            val issueValidator = IssueValidator(issue)
            if (!issueValidator.valid()) {
                val errorMessage = "Issue $jiraId is not valid"
                ErrorNotification.show(errorMessage)
                return
            }
            val commitMessage = CommitMessage(issue)
            commitPanel.setCommitMessage(commitMessage.toString())
        } catch (e: Exception) {
            val errorMessage = "Could not connect to Jira or issue does not exist. Check if $jiraUrl is available and the access token is valid."
            ErrorNotification.show(errorMessage)
        }
    }

    private fun getJiraIdFromBranchName(project: Project): String {
        val repositoryManager = GitUtil.getRepositoryManager(project)
        val branch = repositoryManager.repositories[0].currentBranch ?: return ""
        return branch.name
    }

    companion object {
        private fun getCommitPanel(actionEvent: AnActionEvent?): CommitMessageI? {
            if (actionEvent == null) {
                return null
            }
            val data = Refreshable.PANEL_KEY.getData(actionEvent.dataContext)

            return if (data is CommitMessageI) {
                data
            } else VcsDataKeys.COMMIT_MESSAGE_CONTROL.getData(actionEvent.dataContext)
        }
    }
}
